src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs

```diff
@@ -2218,6 +2218,31 @@ public static bool IsIsOrAsContext(this SyntaxTree syntaxTree, int position, Syn
             // is/as are valid after expressions.
             if (token.IsLastTokenOfNode<ExpressionSyntax>())
             {
+                var expression = token.GetAncestor<ExpressionSyntax>();
```

Comment:
:memo: This is a bit scary, considering how expressions can be nested. Consider the following:

```
((Action)(() => { foreach $$ }))();
```

Even though the `$$` position has an ancestor expression, it's certainly not the one we'd want to use for determining the validity of `is`/`as` at that location. However, it's unclear whether this would pose a problem.

Reply:
I wouldn't expect your example to be a problem because of the `token.IsLastTokenOfNode<ExpressionSyntax>()` check. It's worth a test though.